### PR TITLE
[InstCombine] Add folds for (add/sub/disjoint_or/icmp C, (ctpop (not x)))

### PR DESCRIPTION
**Pull request by goldsteinn**
- [InstCombine] Add tests for folding `(add/sub/disjoint_or/icmp C, (ctpop (not x)))`; NFC
- [InstCombine] Add folds for `(add/sub/disjoint_or/icmp C, (ctpop (not x)))`
----
`(ctpop (not x))` <-> `(sub nuw nsw BitWidth(x), (ctpop x))`. The
`sub` expression can sometimes be constant folded depending on the use
case of `(ctpop (not x))`.

This patch adds fold for the following cases:

`(add/sub/disjoint_or C, (ctpop (not x))`
    -> `(add/sub/disjoint_or C', (ctpop x))`
`(cmp pred C, (ctpop (not x))`
    -> `(cmp swapped_pred C', (ctpop x))`

Where `C'` depends on how we constant fold `C` with `BitWidth(x)` for
the given opcode.

Proofs: https://alive2.llvm.org/ce/z/qUgfF3

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 1]
```diff
@@ -740,6 +740,83 @@ static Value *tryFactorization(BinaryOperator &I, const SimplifyQuery &SQ,
...
+  case Instruction::Add:
+  case Instruction::ICmp:
+    break;
+  }
+  // Find ctpop.
+  auto *Ctpop = dyn_cast<IntrinsicInst>(I->getOperand(1 - ConstIdx));
+  if (Ctpop == nullptr)
+    return nullptr;
+  if (Ctpop->getIntrinsicID() != Intrinsic::ctpop)
+    return nullptr;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:776]:
```suggestion
  Value *Op;
  if (!match(I->getOperand(1 - ConstIdx), m_OneUse(m_Intrinsic<Intrinsic::ctpop>(m_Value(X)))))
    return nullptr;
```

Missing one-use constraints for ctpop?

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp]: Fixed + added test.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 2]
```diff
@@ -740,6 +740,87 @@ static Value *tryFactorization(BinaryOperator &I, const SimplifyQuery &SQ,
...
+  switch (Opc) {
+  case Instruction::Sub:
+    R = Builder.CreateAdd(CtpopOfNotOp, ConstantExpr::getSub(C, BitWidthC));
+    break;
+  case Instruction::Or:
+  case Instruction::Add:
+    R = Builder.CreateSub(ConstantExpr::getAdd(C, BitWidthC), CtpopOfNotOp);
+    break;
+  case Instruction::ICmp:
+    R = Builder.CreateICmp(cast<ICmpInst>(I)->getSwappedPredicate(),
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:814]: Looks like it doesn't hold for `ctpop.i2`.
Alive2: https://alive2.llvm.org/ce/z/CiYgW4

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:814]: Ah, good catch. The lemma also requires C is positive.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:814]: i just disabled to signed predicates. Shouldn't make a difference as any comparison of `ctpop, C` should be trivial to simplify in unsigned form.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for approval from @nikic.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 3]
```diff
@@ -740,6 +740,93 @@ static Value *tryFactorization(BinaryOperator &I, const SimplifyQuery &SQ,
...
+    return nullptr;
+
+  Constant *C;
+  // Check other operand is ImmConstant.
+  if (!match(I->getOperand(ConstIdx), m_ImmConstant(C)))
+    return nullptr;
+
+  Type *Ty = Op->getType();
+  Constant *BitWidthC = ConstantInt::get(Ty, Ty->getScalarSizeInBits());
+  // Need extra check for icmp. Note if this check is it generally means the
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:791]: I think something is missing after "if this check is"?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**